File: DICCIONARIOS/Python/9.py
# Gestion de facturas pendientes de cobro 

facturas = {}    # Diccionario vacío
cobrado = 0      # Total acumulado ya cobrado

def pendiente_total(dic):
    # Funcion que devuelve la suma de todos los valores del diccionario (lo pendiente a pagar)
    return sum(dic.values())

while True:    # Bucle principal, se repite hasta que el usuario decida terminar
    print("\nOpciones: (A)ñadir  (P)agar  (T)erminar")
    op = input("Elija una opcion: ").strip() # Leer opcion del usuario y pasarla a mayusculas

    if op == "A":   # Si elige añadir factura
        num = input("Numero de factura: ").strip()    # Pedimos el numero de factura
        if num in facturas:                          # Si ya existe no se agrega
            print("Esa factura ya existe.")
        else:
            try:
                costo = int(input("Costo de la factura: ").strip()) # Pedimos el costo 
                facturas[num] = costo                              # Guardamos en el diccionario
                print(f"Factura {num} añadida por {costo}.")
            except ValueError:   # Si escribe algo que no es número
                print("Costo invalido. No se añadio la factura.")

    elif op == "P":   # Si elige pagar factura
        num = input("Numero de factura a pagar: ").strip()  # Pedimos el numero
        if num in facturas:                                 # Si existe en el diccionario
            importe = facturas.pop(num)   # Eliminamos la factura y guardamos su costo
            cobrado += importe            # Sumamos al total cobrado
            print(f"Factura {num} pagada por {importe}.")
        else:   # Si no existe
            print("No existe esa factura.")

    elif op == "T":   # Si elige terminar
        break          # Se rompe el bucle y el programa acaba
    else:
        print("Opcion no valida.")   # Si escribe cualquier cosa diferente
        continue

    # Mostrar totales después de cada operación válida
    pendiente = pendiente_total(facturas)
    print(f"Cobrado hasta el momento: {cobrado}")
    print(f"Pendiente de cobro:       {pendiente}")

# Al salir mostramos el resumen final
print("\nPrograma terminado.")
print(f"Total cobrado: {cobrado} | Pendiente final: {pendiente_total(facturas)}")

File: DICCIONARIOS/Java/9.java
package ejercicio5;

import java.util.HashMap;   // Importamos HashMap, estructura clave-valor (diccionario)
import java.util.Map;       // Interfaz general Map
import java.util.Scanner;   // Para leer datos del usuario desde consola

public class Facturas {
	
	 public static void main(String[] args) {
	        Scanner sc = new Scanner(System.in); // Objeto Scanner para leer del teclado
	        Map<String, Integer> facturas = new HashMap<>();  // Diccionario: clave=numero, valor=costo (int)
	        int cobrado = 0;  // Total ya cobrado

	        while (true) {   // Bucle principal
	            System.out.println("\nOpciones: 1. Añadir  2. Pagar  3. Terminar");
	            System.out.print("Elija una opcion: ");
	            String op = sc.nextLine().trim();   // se lee la opcion

	            switch (op) {  // Se evalua la opcion ingresada
	                case "1": {   // Añadir factura
	                    System.out.print("Numero de factura: "); // Leemos el numero de factura
	                    String numero = sc.nextLine().trim(); 

	                    if (facturas.containsKey(numero)) {
	                        System.out.println("Esa factura ya existe.");  // Si ya existe no se la añade
	                    } else {
	                        System.out.print("Costo de la factura: ");
	                        try {
	                            int costo = Integer.parseInt(sc.nextLine().trim());
	                            facturas.put(numero, costo); // Guardamos en el diccionario
	                            System.out.println("Factura " + numero + " añadida por " + costo + ".");
	                        } catch (NumberFormatException e) {   // Si el valor no es numero entero
	                            System.out.println("Costo invalido. No se añadio la factura.");
	                        }
	                    }
	                    break;
	                }
	                case "2": {   // Pagar factura
	                    System.out.print("Numero de factura a pagar: ");
	                    String numero = sc.nextLine().trim();

	                    Integer importe = facturas.remove(numero);  // Eliminamos la factura y obtenemos su valor
	                    if (importe != null) {   // Si la factura existia
	                        cobrado += importe;  // Sumamos al total cobrado
	                        System.out.println("Factura " + numero + " pagada por " + importe + ".");
	                    } else {
	                        System.out.println("No existe esa factura.");
	                    }
	                    break;
	                }
	                case "3":    // Terminar el programa
	                    int pendienteFinal = sumaValores(facturas);  // se calcula lo pendiente
	                    System.out.println("\nTotal cobrado: " + cobrado +
	                                       " | Pendiente final: " + pendienteFinal);
	                    sc.close();   // Cerramos el Scanner
	                    return;
	                default:  // Si la opcion no es valida
	                    System.out.println("Opcion no valida.");
	                    continue; // vuelve a mostrar el menu
	            }

	            // Mostrar siempre los totales despues de cada operacion valida
	            int pendiente = sumaValores(facturas);
	            System.out.println("Cobrado hasta el momento: " + cobrado);
	            System.out.println("Pendiente de cobro:       " + pendiente);
	        }
	    }

	  // Funcion auxiliar que suma todos los valores del diccionario
	    private static int sumaValores(Map<String, Integer> mapa) {
	        int s = 0;
	        for (int v : mapa.values()) s += v;  // Recorremos todos los valores y sumamos
	        return s;
	    }
}
